CODE_STYLE.md

```diff
@@ -0,0 +1,943 @@
+
+  - Please **include the appropriate test cases** for your patch.
+
+  - Make sure all tests pass before submitting your changes. See the
+    [Makefile operations](/README.md#makefile-operations).
+
+  - Make sure the linter does not throw any errors: `make lint`.
+
+  - Rebase your commits. It may be that new commits have been introduced on `next`. Rebasing
+    will update your branch with the most recent code and make your changes easier to review:
+
+    ```
+    $ git fetch
+    $ git rebase origin/next
+    ```
+
+  - Push your changes:
+
+    ```
+    $ git push origin -u feature/my-feature
+    ```
+
+  - Open a pull request against the [appropriate branch][#branches].
+
+  - If we suggest changes:
+  
+    - Please make the required updates (after discussion if any)
+    - Re-run the test suite
+    - Only create new commits if it makes sense. Generally, you will want to 
+      [amend your latest commit or rebase your branch](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html)
+      after the new changes:
+
+      ```
+      $ git rebase -i next
+      # choose which commits to edit and perform the updates
+      ```
+
+    - Force push to your branch:
+
+      ```
+      $ git push origin feature/my-feature -f
+      ```
+  
+    **[[back to top]](#TOC)**
+
+## <a name='testing'>Testing</a>
+
+  - Use [busted](http://olivinelabs.com/busted) and write lots of tests in a /spec 
+    folder. Separate tests by module.
+  - Use descriptive `describe` and `it` blocks so it's obvious to see what
+    precisely is failing.
+  - Test one functionality per test, repeat for multiple functionalities.
+  - Tests should be atomic, meaning they can only rely on `setup` and `before_each`
+    functionalities, but never on the results or changes by a previous test (even
+    if run on its own it should pass).
+
+  - For the keyword assertions (`nil`, `true`, and `false`) use the `is_xxx` format
+
+    ```lua
+    --bad
+    assert.Nil(something)
+
+    --good
+    assert.is_nil(something)
+    ```
+
+  - Use specific assertions where available. The specific assertion will provide better
+    error messages when they fail, as they better understand the context.
+
+    ```lua
+    --bad
+    local value = r.body.headers["x-something"]
+    assert(value == "something","expected 'something'")
+
+    --good
+    local value = assert.request(r).has.header("x-something")
+    assert.equal("something", value)
+    ```
+
+    **[[back to top]](#TOC)**
+
+## <a name='performance'>Performance</a>
+
+  - cache globals when used repeatedly. Locals are faster than looking up globals.
+
+    ```lua
+    --bad
+    for i = 1, 20 do
+      t[i] = math.floor(t[i])
+    end
+
+    --good
+    local math_floor = math.floor
+    for i = 1, 20 do
+      t[i] = math_floor(t[i])
+    end
+    ```
+
+  - Insertion into lists by using the length operator `#`. Never use the 'length' 
```

Comment:
Performance should definitely include notes about pre-allocated tables as well as JIT-able code.